programs/pluto/ikev2_ts.c

```diff
@@ -340,13 +504,55 @@ static bool v2_parse_ts(struct payload_digest *const ts_pd,
 		return false;	/* won't fit in array */
 	}
 
+	/*
+	 * addr_range_count: Number of address range substructures in the
+	 * Traffic Selector (TS) payload
+	 */
+	uint32_t addr_range_count = 0;
+	/*
+	 * seclabel_count: Number of security label substructures in the TS
+	 * payload.
+	 * NOTE: Currently, only 1 TS_SECLABEL substructure is expected in the
+	 * TS Payload.
+	 */
+	uint32_t seclabel_count = 0;
+	/*
+	 * seclabel_index: Index of the TS_SECLABEL substructure in the array of Traffic
+	 * Selectors corresponding to the TS Payload.
+	 * NOTE: Currently, only 1 TS_SECLABEL substructure is expected in the
+	 * TS Payload.
+	 */
+	int seclabel_index = -1;
 	for (tss->nr = 0; tss->nr < ts_pd->payload.v2ts.isat_num; tss->nr++) {
 		struct traffic_selector *ts = &tss->ts[tss->nr];
+		/* Zero-out the Traffic Selector to prevent garbage data. */
+		zero(ts);
+
+		/* ts_type: Traffic Selector Type */
+		uint8_t ts_type;
+		/*
+		 * Peek at the Traffic Selector Type in the input stream in order to figure
+		 * out which substructure to process next.
+		 */
+		if (!peek_raw(&ts_type, sizeof(ts_type), &ts_pd->pbs, "Traffic Selector Type")) {
+			return false;
+		}
+
+		if (ts_type == IKEv2_TS_SECLABEL) {
```

Comment:
this could have been checked as an additional case in the switch below (along with v4 and v6 traffic types)

Reply:
I did have it as an additional case in the `switch` initially, but that resulted in duplicating calls to `in_struct()` for `&ts1` and `++addr_range_count`. What I ended up having was the cleanest code I could write.

But let me know if I should still add a case to the `switch`.